### PR TITLE
fix: avoid empty fast bitset memcpy

### LINKED ISSUE #1963: Fix FastBitset zero-length memcpy with null data
Parent tracking issue: #1960

The x86 ASan/UBSan jobs report a null pointer passed to `memcpy` in `FastBitset::resize`.

CI evidence:

- Test X86 Unittests: https://github.com/antgroup/vsag/actions/runs/25100797892/job/73553696204
- Test X86 Functests: https://github.com/antgroup/vsag/actions/runs/25100797892/job/73553696201

Log excerpts:

```text
/__w/vsag/vsag/src/impl/bitset/fast_bitset.cpp:237:20: runtime error: null pointer passed as argument 2, which is declared to never be null
```

Code location:

- `src/impl/bitset/fast_bitset.cpp:237`: `std::memcpy(tmp, data_, size_ * sizeof(uint64_t));`

Initial root cause:

When a `FastBitset` has `size_ == 0` and `data_ == nullptr`, resizing to a larger capacity calls `memcpy` with a zero byte count but a null source pointer. UBSan still reports this as invalid because the C library declaration marks the argument non-null.

Suggested fix:

- In `FastBitset::resize`, call `std::memcpy` only when `size_ > 0`.
- Ensure zero-size deserialize/resize paths leave the object in a valid empty state.
- Add a regression test for resizing an empty `FastBitset` and, if relevant, deserializing an empty bitset.

Non-goals:

- Do not change bitset semantics or public API behavior.
- Do not mix HNSW or other UBSan fixes into this PR.

## Automated CI failure occurrences

Observed **1** time. This table is maintained by @vsag-bot.

| Time (UTC) | Branch | Commit | Workflow | Failure |
| --- | --- | --- | --- | --- |
| 2026-09-15T13:17:55.000Z | 0.18 (0.18) | [`b07774e`](https://github.com/antgroup/vsag/commit/b07774e24b4b1b07a2ddbab5736c0b9d2f009712) | Asan Build & Test Parallel | [run #34944183382 attempt 1](https://github.com/antgroup/vsag/actions/runs/34944183382) |

### PR DESCRIPTION
**Pull request by jac0626**
## Summary

Fix `FastBitset` zero-length serialization edge cases that can pass null pointers to `memcpy` under UBSan.

The deserialize path now keeps empty bitsets represented without allocating data, and resize only copies existing words when the previous size is non-zero. A regression test covers resizing from an empty bitset and deserializing an empty bitset.

Fixes #1963.
Part of #1960.

## Validation

- `git diff --check`
- Fork PR CI: https://github.com/jac0626/vsag/actions/runs/25476854514
- Log scan: target `fast_bitset.cpp` runtime error is gone in this branch; remaining runtime errors are existing issues tracked separately (`inner_index_interface.cpp`, `sparse_term_datacell.cpp`, `hnswalg.*`).

## Notes

Local macOS rebuild was blocked by the existing CMake configuration error: `gcc does not support using libc++`.

### CODE REVIEW

File: src/impl/bitset/fast_bitset.cpp [hunk 1]
```diff
@@ -217,8 +217,12 @@ FastBitset::Deserialize(StreamReader& reader) {
...
-    reader.Read(reinterpret_cast<char*>(data_), size * sizeof(uint64_t));
+    delete[] data_;
+    data_ = nullptr;
+    if (size > 0) {
+        data_ = new uint64_t[size];
+        reader.Read(reinterpret_cast<char*>(data_), size * sizeof(uint64_t));
+    }
     this->size_ = size;
     this->set_capacity(size);
     this->set_fill_bit(fill_bit);
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/impl/bitset/fast_bitset.cpp:249]: Fixed in `69fc4673`. `Deserialize` now validates the serialized word count against the 31-bit capacity limit before allocation, reads into temporary storage, and only replaces the current object state after the read succeeds. Added a regression test that rejects an oversized serialized size and verifies the existing bitset remains intact.

File: src/impl/bitset/fast_bitset.cpp [hunk 2]
```diff
@@ -217,10 +220,20 @@ FastBitset::Deserialize(StreamReader& reader) {
...
-    this->set_capacity(size);
+    if (size > MAX_FAST_BITSET_WORDS) {
+        throw VsagException(ErrorType::INTERNAL_ERROR, "bitset size too large");
+    }
+
+    std::unique_ptr<uint64_t[]> new_data;
+    if (size > 0) {
+        new_data = std::make_unique<uint64_t[]>(size);
+        reader.Read(reinterpret_cast<char*>(new_data.get()), size * sizeof(uint64_t));
+    }
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/impl/bitset/fast_bitset.cpp:242]: Fixed in `d1324e43`. Added an explicit `size <= std::numeric_limits<size_t>::max() / sizeof(uint64_t)` guard before allocation/read.

File: src/impl/bitset/fast_bitset.cpp [hunk 3]
```diff
@@ -217,10 +220,20 @@ FastBitset::Deserialize(StreamReader& reader) {
     StreamReader::ReadObj(reader, fill_bit);
     uint64_t size;
     StreamReader::ReadObj(reader, size);
-    data_ = new uint64_t[size];
-    reader.Read(reinterpret_cast<char*>(data_), size * sizeof(uint64_t));
-    this->size_ = size;
-    this->set_capacity(size);
+    if (size > MAX_FAST_BITSET_WORDS) {
+        throw VsagException(ErrorType::INTERNAL_ERROR, "bitset size too large");
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/impl/bitset/fast_bitset.cpp]: Fixed in `d1324e43`. Oversized serialized input now throws `ErrorType::READ_ERROR` instead of `INTERNAL_ERROR`.

File: src/impl/bitset/fast_bitset.cpp [hunk 4]
```diff
@@ -15,12 +15,15 @@
...
 
+#include <memory>
+
 #include "simd/bit_simd.h"
 #include "vsag_exception.h"
 
 namespace vsag {
 
 static constexpr uint64_t FILL_ONE = 0xFFFFFFFFFFFFFFFF;
+static constexpr uint64_t MAX_FAST_BITSET_WORDS = 0x7FFFFFFF;
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/impl/bitset/fast_bitset.cpp]: Fixed in `d1324e43`. Replaced the hex literal with a limit derived from `std::numeric_limits<uint32_t>::max() >> 1`, matching the capacity bit layout.

File: src/impl/bitset/fast_bitset.cpp [hunk 5]
```diff
@@ -15,12 +15,16 @@
...
 
+#include <limits>
+#include <memory>
+
 #include "simd/bit_simd.h"
 #include "vsag_exception.h"
 
 namespace vsag {
 
 static constexpr uint64_t FILL_ONE = 0xFFFFFFFFFFFFFFFF;
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/impl/bitset/fast_bitset.cpp:27]: Fixed in `b6ba7f50`. Added an in-code comment explaining that the high bit of `capacity_` stores `fill_bit`, so the serialized word count must fit in 31 bits.

File: src/impl/bitset/fast_bitset.cpp [hunk 6]
```diff
@@ -217,10 +221,23 @@ FastBitset::Deserialize(StreamReader& reader) {
     StreamReader::ReadObj(reader, fill_bit);
     uint64_t size;
     StreamReader::ReadObj(reader, size);
-    data_ = new uint64_t[size];
-    reader.Read(reinterpret_cast<char*>(data_), size * sizeof(uint64_t));
-    this->size_ = size;
-    this->set_capacity(size);
+    if (size > MAX_FAST_BITSET_WORDS) {
+        throw VsagException(ErrorType::READ_ERROR, "bitset size too large");
+    }
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/impl/bitset/fast_bitset.cpp:230]: Fixed in `b6ba7f50`. This duplicated comment is covered by the new capacity layout comment near `MAX_FAST_BITSET_WORDS`.

File: src/impl/bitset/fast_bitset.cpp [hunk 7]
```diff
@@ -217,10 +221,23 @@ FastBitset::Deserialize(StreamReader& reader) {
...
+    if (size > std::numeric_limits<size_t>::max() / sizeof(uint64_t)) {
+        throw VsagException(ErrorType::READ_ERROR, "bitset byte size too large");
+    }
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/impl/bitset/fast_bitset.cpp:236]: Fixed in `b6ba7f50`. The read errors now include the offending serialized word count and the relevant maximum, with messages distinguishing word-size and byte-size limits.

File: src/impl/bitset/fast_bitset_test.cpp [hunk 8]
```diff
@@ -199,6 +201,50 @@ GetUnion(const std::vector<int>& values1, const std::vector<int>& values2) {
...
 }
 
+TEST_CASE("FastBitset empty resize and deserialize", "[ut][FastBitset]") {
+    auto allocator = SafeAllocator::FactoryDefaultAllocator();
+
+    FastBitset bitset(allocator.get());
+    bitset.Set(3, true);
+    REQUIRE(bitset.Test(3));
+    REQUIRE(bitset.Count() == 1);
+
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/impl/bitset/fast_bitset_test.cpp]: Fixed in `b6ba7f50`. Removed the unrelated non-empty bitset setup from the empty deserialize regression test.

**Review by LHT129 — APPROVED** [review]
LGTM
